app/controllers/analysis_controller.rb

```diff
@@ -0,0 +1,420 @@
+      # Obtain death year of decedent.
+      record_death_year = DateTime.parse(record.contents['datePronouncedDead.datePronouncedDead']).year if record.contents['datePronouncedDead.datePronouncedDead']
+      # Obtain death month of decedent.
+      record_death_month = DateTime.parse(record.contents['datePronouncedDead.datePronouncedDead']).month if record.contents['datePronouncedDead.datePronouncedDead']
+      # Obtain death day of decedent.
+      record_death_day = DateTime.parse(record.contents['datePronouncedDead.datePronouncedDead']).day if record.contents['datePronouncedDead.datePronouncedDead']
+      # Obtain sex of decedent.
+      record_sex = record.contents['sex.sex'] if record.contents['sex.sex']
+      # Obtain races of the decedent.
+      record_races_listed = JSON.parse(record.contents["race.race.specify"]) if record.contents["race.race.specify"]
+      next if (record_birth_year.nil? || record_death_year.nil?) && decedent_age_range.length > 0
+      age_at_death = record_death_year - record_birth_year
+      # Adjust age at death based upon time of birth and time of death.
+      if record_death_month < record_birth_month
+        age_at_death -= 1
+      elsif record_death_month == record_birth_month
+        if record_death_day < record_birth_day
+          age_at_death -= 1
+        end
+      end
+      # Skip decedent if their gender was not specified and user asked to filter based on gender.
+      next if record_sex.nil? && decedent_sex.length > 0
+      # Skip decedent if their death year is before the start year specified or if their death year is after the end year specified.
+      next if record_death_year < start_year || record_death_year > end_year
+      # Skip if age of decedent is not within any of the age ranges specified by user.
+      next if partial_age_hash.select{|a| a === age_at_death}.empty? && !partial_age_hash.empty?
+      # Skip if decedent has no recorded race and user asked to filter based on race.
+      next if !decedent_race.empty? && record_races_listed.nil?
+      # Skip if decedent has no races in common with the races specified by the user.
+      next if (!decedent_race.empty? && (record_races_listed & decedent_race).empty?)
+      target_decedent_cods << record.contents['cod.immediate'] if record.contents['cod.immediate'].present?
+    end
+    counts = target_decedent_cods.each_with_object(Hash.new(0)) { |word,counts| counts[word] += 1 }
+    hash_sent = {}
+    num_counted = 0
+    # Sort the hash in reverse order from highest cause of death to lowest and store the entries in the hash_sent hash.
+    counts.sort_by {|k,v| v}.reverse.each do |key,val|
+      hash_sent[key] = val
+      num_counted += 1
+      if num_counted == num_causes
+        break
+      end
+    end
+    render json: hash_sent
+  end
+
+  # Function which will return chart information to front end.
+  def analyze_causes
```

Comment:
in one case we're trying to find the number of people who passed away in a particular year from a specified cause of death. the other is trying to find the top n causes of death